src/main/java/org/broadinstitute/hellbender/tools/exome/cnlohcaller/CNLOHCaller.java

```diff
@@ -0,0 +1,803 @@
+    }
+
+    /**
+     * Make both CNLoH and balanced segment calls.
+     *
+     * This method is not thread-safe, since it initializes member attributes.
+     *
+     * @param segments The segments with segment mean and minor allelic fraction estimates (typically from ACNV).
+     * @param numIterations Number of EM iteration to run
+     * @param ctx This call requires Spark, so not {@code null}
+     * @return Never {@code null}
+     */
+    public List<CNLOHCall> makeCalls(final List<ACNVModeledSegment> segments, final int numIterations, final JavaSparkContext ctx) {
+
+        ParamUtils.isPositive(numIterations, "Must be more than zero iterations.");
+        Utils.nonNull(segments);
+        Utils.nonNull(ctx, "Java SparkContext can't be null when attempting to make CNLOH and balance calls.");
+
+        segmentMeanBiasInCR = calculateSegmentMeanBiasInCRSpace(segments);
+        segmentMeanVarianceInCR = calculateVarianceOfCopyNeutralSegmentMeans(segments, segmentMeanBiasInCR);
+
+        CNLOHCallerModelState state = CNLOHCallerModelState.createInitialCNLOHCallerModelState(rhoThreshold, segments, normalNumCopies);
+
+        // Create a Spark RDD for the segments.
+        final JavaRDD<ACNVModeledSegment> segs = ctx.parallelize(segments);
+
+        List<CNLOHCall> cnlohCalls = null;
+
+        // for each iteration
+        for (int i = 1; i <= numIterations; i++) {
+
+            logger.info("Starting iteration " + i + " of " + numIterations + " ... ");
+            logger.info("Calculating big E_zsk_vsm_wsn for each segment (iteration " + i + " of " + numIterations + ") ... ");
+
+            // We are not broadcasting state, since it will only be used once before values change.  In that case,
+            //  there is no gain for broadcasting
+
+            // Effectively, the responsibilities are a 4D array:  S x K x M x N.  We do the responsibility calculation
+            //  over the segments.  So the JavaRDD will be of length S.
+            // Update E_zsk_vsm_wsn (responsibilities) for each segment (Steps 1 & 2)  {K x M x N} in S entries.  The RDD is of length S.
+            final JavaRDD<double[][][]> eZskVsmWsnBySeg = segs.map(s -> calcE_zsk_vsm_wsn(state.geteLnPhiK(), state.getCnToPiMap(), state.getRhos(),
+                    s.getMinorAlleleFractionPosteriorSummary().getCenter(),
+                    s.getMinorAlleleFractionPosteriorSummary().getLower(),
+                    s.getMinorAlleleFractionPosteriorSummary().getUpper(),
+                    Math.pow(2, s.getSegmentMeanPosteriorSummary().getCenter()) - segmentMeanBiasInCR,
+                    Math.pow(2, s.getSegmentMeanPosteriorSummary().getLower()) - segmentMeanBiasInCR,
+                    Math.pow(2, s.getSegmentMeanPosteriorSummary().getUpper()) - segmentMeanBiasInCR,
+                    state.getLambda()));
+
+            // Create an array (for each segment) that is a sum of eZskVsmWsnBySeg on the rhos.  (length:K for S entries)
```

Comment:
Are you sure this is a sum on the rhos?  It looks like a sum on the m's and n's.